File: packages/core_dart/test/extract_accent_test.dart
import 'package:kromia_core/kromia_core.dart';
import 'package:test/test.dart';

/// Espejo 1:1 del corpus TS `tests/extract-accent.test.ts` (KRO-73).
/// Misma entrada → misma salida que `@kromia/core` (paridad vigilada por drift-CI).
void main() {
  const colorField = FieldDefLike(key: 'color', type: 'text', behavior: 'color_hex');

  ViewComposition comp(Map<String, dynamic> json) =>
      ViewComposition.fromJson(<String, dynamic>{'recipe': 'compact_avatar', ...json});

  group('extractAccentSettings — color hex detection', () {
    test('item con color hex válido (#ff0000) → color + position', () {
      final r = extractAccentSettings(null, {'color': '#ff0000'}, [colorField], 'top');
      expect(r, const AccentSettings(color: '#ff0000', position: 'top'));
    });

    test('color hex sin "#" (ff0000) → añade el prefix', () {
      final r = extractAccentSettings(null, {'color': 'ff0000'}, [colorField], 'top');
      expect(r?.color, '#ff0000');
    });

    test('color hex con alpha (#ff000080) → válido', () {
      final r = extractAccentSettings(null, {'color': '#ff000080'}, [colorField], 'top');
      expect(r?.color, '#ff000080');
    });

    test('color hex inválido (xyz) → null', () {
      final r = extractAccentSettings(null, {'color': 'xyz'}, [colorField], 'top');
      expect(r, isNull);
    });

    test('sin fields color_hex → null', () {
      final r = extractAccentSettings(
          null, {'nombre': '#ff0000'}, [const FieldDefLike(key: 'nombre', type: 'text')], 'top');
      expect(r, isNull);
    });

    test('item sin valor en el field color → null', () {
      final r = extractAccentSettings(null, {}, [colorField], 'top');
      expect(r, isNull);
    });

    test('múltiples color_hex → coge el primero', () {
      final r = extractAccentSettings(
        null,
        {'c1': '#ff0000', 'c2': '#00ff00'},
        [
          const FieldDefLike(key: 'c1', type: 'text', behavior: 'color_hex'),
          const FieldDefLike(key: 'c2', type: 'text', behavior: 'color_hex'),
        ],
        'top',
      );
      expect(r?.color, '#ff0000');
    });
  });

  group('extractAccentSettings — resolución de posición', () {
    test('sin overrides → recipeDefault', () {
      final r = extractAccentSettings(null, {'color': '#ff0000'}, [colorField], 'left');
      expect(r?.position, 'left');
    });

    test('composition.accentPosition="top" → "top"', () {
      final r = extractAccentSettings(
          comp({'slots': {}, 'accentPosition': 'top'}), {'color': '#ff0000'}, [colorField], 'left');
      expect(r?.position, 'top');
    });

    test('composition.accentPosition="none" → "none" (NO cae a default)', () {
      final r = extractAccentSettings(
          comp({'slots': {}, 'accentPosition': 'none'}), {'color': '#ff0000'}, [colorField], 'top');
      expect(r?.position, 'none');
    });

    test('composition "auto" + slot.appearance override → slot override', () {
      final r = extractAccentSettings(
        comp({
          'slots': {
            'title': {
              'fields': ['color'],
              'appearance': {'accentPosition': 'right'},
            },
          },
          'accentPosition': 'auto',
        }),
        {'color': '#ff0000'},
        [colorField],
        'top',
      );
      expect(r?.position, 'right');
    });

    test('composition "auto" + slot sin override → recipeDefault', () {
      final r = extractAccentSettings(
        comp({
          'slots': {
            'title': {'fields': ['color']},
          },
          'accentPosition': 'auto',
        }),
        {'color': '#ff0000'},
        [colorField],
        'left',
      );
      expect(r?.position, 'left');
    });

    test('composition explícito > slot override (composition gana)', () {
      final r = extractAccentSettings(
        comp({
          'slots': {
            'title': {
              'fields': ['color'],
              'appearance': {'accentPosition': 'right'},
            },
          },
          'accentPosition': 'bottom',
        }),
        {'color': '#ff0000'},
        [colorField],
        'top',
      );
      expect(r?.position, 'bottom');
    });
  });
}

File: packages/core/tests/extract-accent.test.ts
/**
 * Tests de `extractAccentSettings` — KRO-73 (B+).
 *
 * Cubre cascada de resolución (composition > slot > recipeDefault) +
 * detección de color hex válido.
 */

import { describe, it, expect } from 'vitest';
import { extractAccentSettings } from '../src/extract-accent';
import type { FieldDefLike, SlotComposition } from '../src/types';

const colorField: FieldDefLike = { key: 'color', type: 'text', behavior: 'color_hex' };

describe('extractAccentSettings — color hex detection', () => {
  it('item con color hex válido (#ff0000) → devuelve color + position', () => {
    const r = extractAccentSettings(
      undefined,
      { color: '#ff0000' },
      [colorField],
      'top',
    );
    expect(r).toEqual({ color: '#ff0000', position: 'top', colorFieldKey: 'color' });
  });

  it('expone colorFieldKey del campo que alimenta el acento (KRO-198)', () => {
    const r = extractAccentSettings(
      undefined,
      { c1: '#ff0000', c2: '#00ff00' },
      [
        { key: 'c1', type: 'text', behavior: 'color_hex' },
        { key: 'c2', type: 'text', behavior: 'color_hex' },
      ],
      'top',
    );
    // El primero gana → su key es la que el host suprime como celda.
    expect(r?.colorFieldKey).toBe('c1');
  });

  it('color hex sin "#" (ff0000) → añade el prefix', () => {
    const r = extractAccentSettings(
      undefined,
      { color: 'ff0000' },
      [colorField],
      'top',
    );
    expect(r?.color).toBe('#ff0000');
  });

  it('color hex con alpha (#ff000080) → válido', () => {
    const r = extractAccentSettings(
      undefined,
      { color: '#ff000080' },
      [colorField],
      'top',
    );
    expect(r?.color).toBe('#ff000080');
  });

  it('color hex inválido (xyz) → undefined', () => {
    const r = extractAccentSettings(
      undefined,
      { color: 'xyz' },
      [colorField],
      'top',
    );
    expect(r).toBeUndefined();
  });

  it('sin fields color_hex → undefined', () => {
    const r = extractAccentSettings(
      undefined,
      { nombre: '#ff0000' },
      [{ key: 'nombre', type: 'text' }],
      'top',
    );
    expect(r).toBeUndefined();
  });

  it('item sin valor en el field color → undefined', () => {
    const r = extractAccentSettings(
      undefined,
      {},
      [colorField],
      'top',
    );
    expect(r).toBeUndefined();
  });

  it('múltiples color_hex → coge el primero', () => {
    const r = extractAccentSettings(
      undefined,
      { c1: '#ff0000', c2: '#00ff00' },
      [
        { key: 'c1', type: 'text', behavior: 'color_hex' },
        { key: 'c2', type: 'text', behavior: 'color_hex' },
      ],
      'top',
    );
    expect(r?.color).toBe('#ff0000');
  });
});

describe('extractAccentSettings — resolución de posición', () => {
  it('sin overrides → recipeDefault', () => {
    const r = extractAccentSettings(
      undefined,
      { color: '#ff0000' },
      [colorField],
      'left',
    );
    expect(r?.position).toBe('left');
  });

  it('composition.accentPosition="top" → "top"', () => {
    const r = extractAccentSettings(
      { slots: {}, accentPosition: 'top' },
      { color: '#ff0000' },
      [colorField],
      'left',
    );
    expect(r?.position).toBe('top');
  });

  it('composition.accentPosition="none" → "none" (NO cae a default)', () => {
    const r = extractAccentSettings(
      { slots: {}, accentPosition: 'none' },
      { color: '#ff0000' },
      [colorField],
      'top',
    );
    expect(r?.position).toBe('none');
  });

  it('composition.accentPosition="auto" + slot.appearance override → slot override', () => {
    const slots: Record<string, SlotComposition> = {
      title: {
        fields: ['color'],
        appearance: { accentPosition: 'right' },
      },
    };
    const r = extractAccentSettings(
      { slots, accentPosition: 'auto' },
      { color: '#ff0000' },
      [colorField],
      'top',
    );
    expect(r?.position).toBe('right');
  });

  it('composition.accentPosition="auto" + slot sin override → recipeDefault', () => {
    const slots: Record<string, SlotComposition> = {
      title: { fields: ['color'] },
    };
    const r = extractAccentSettings(
      { slots, accentPosition: 'auto' },
      { color: '#ff0000' },
      [colorField],
      'left',
    );
    expect(r?.position).toBe('left');
  });

  it('composition explícito > slot override (composition gana)', () => {
    const slots: Record<string, SlotComposition> = {
      title: {
        fields: ['color'],
        appearance: { accentPosition: 'right' },
      },
    };
    const r = extractAccentSettings(
      { slots, accentPosition: 'bottom' },
      { color: '#ff0000' },
      [colorField],
      'top',
    );
    expect(r?.position).toBe('bottom');
  });
});
